graph_def_editor/graph.py

```diff
@@ -1001,42 +684,218 @@ def get_frame_names(self):
+      escape_functions=False  # type: bool
+  ):
     # type: (...) -> None
     """
     Visit all nodes reachable from a starting set in the order of a
-    breadth-first traversal. Invokes a callback at each node visited.
+    breadth-first traversal (going from node to output edges).
+    Invokes a callback at each node visited.
 
     Args:
       visitor: Possibly-stateful callback to be invoked on each node reached
       starting_nodes: Optional list of starting nodes. If this set is not
-         provided, this method will use all nodes with zero inputs as the
-         starting set. Search will visit these nodes first, then visit their
-         children in order by parent node.
+          provided, this method will use all nodes with zero inputs as the
+          starting set. Search will visit these nodes first, then visit their
+          children in order by parent node.
+      visited_nodes: Optional set of nodes to skip iterating over.
+      iterate_functions: Indicates if we should also go inside functions if one
+          is found in the graph.
+      escape_functions: If iteration started in a function graph, indicates
+          that we should also iterate through all function callers up
+          the stack.
+    Returns:
+      True if iteration was iterruputed by visitor, otherwise False.
     """
     if starting_nodes is None:
       # Start with all of the nodes in the graph that have no inputs.
       # The maintainers of the TensorFlow scheduler like to call these nodes
       # "root nodes".
-      starting_nodes = [n for n in self.nodes if 0 == len(n.inputs)]
+      starting_nodes = [n for n in self.nodes if not n.inputs]
+
+    if visited_nodes is None:
+      visited_nodes = set()
+
+    nodes_queue = queue.Queue()
+    function_graph_names_set = set()
+
+    starting_nodes_set = set()
+    for n in starting_nodes:
+      nodes_queue.put((n, None))
+      starting_nodes_set.add(n)
+
+    while not nodes_queue.empty():
+      (n, input_tensor) = nodes_queue.get()
+      if n in visited_nodes:
+        continue
+      if n.op_type != _INPUT_DUMMY_OP_NAME:
+        if visitor(n):
+          return True
+
+      if escape_functions and isinstance(n.graph, function_graph.FunctionGraph):
+        function_graph_names_set.add(n.graph.name)
+
+      visited_nodes.add(n)
+      if iterate_functions and n.op_type in node.PARTITIONED_CALL_OP_TYPES:
+        function_name = n.get_attr("f").name
+        f_graph = self.get_function_graph_by_name(function_name)
+
+        function_inputs = []
+        if input_tensor is not None:
+          for input_node in f_graph.nodes:
+            if input_node.op_type == _INPUT_DUMMY_OP_NAME and input_node.name in input_tensor.name:
+              function_inputs.append(input_node)
+
+        if len(function_inputs) == 0:
+          function_inputs = [input for input in f_graph.nodes if input.op_type == _INPUT_DUMMY_OP_NAME]
+
+        if self.breadth_first_visitor(
```

Comment:
This part of the traversal doesn't appear to be breadth-first. Shouldn't this code be emptying `nodes_queue` before recursing into the function body? Or perhaps passing `nodes_queue` down to the recursive call?

Reply:
You are right, strictly speaking it is not BFS, the right way would be to extract all nodes of subgraph, merge with current one and do a traversal. But following such logic, nodes from parent graph might get interleaved with nodes in child graph, and that is undesirable.

From this perspective current logic makes more sense - once we get to the function call, we iterate all nodes in the function in a BFS manner and then continue with the rest of parent graph nodes.

I'll update comment on the method.